website/static/js/registrationUtils.js

```diff
@@ -0,0 +1,826 @@
+            // Note that if a save request has not happened, the form is considered dirty.
+            return true;
+        }
+        if (request.payload === undefined) {
+            // If the payload is not attached, the .done() handler hasn't been called;
+            //      therefore the request has not successfully completed.
+            return false;
+        }
+        return JSON.stringify(request.payload) !== JSON.stringify(self.serialized());
+    });
+
+    /* The last autosave request that was sent to the server. */
+    self.lastSaveRequest = ko.observable();
+    self.lastSaveAge = ko.pureComputed(function () {
+        // TODO(lyndsysimon): Test this.
+        var request = self.lastSaveRequest();
+        if (request === undefined) {
+            // not yet saved
+            return;
+        } else if (request.completedDate === undefined) {
+            // request in flight
+            return -1;
+        } else {
+            return (new Date() - request.completedDate);
+        }
+    });
+
+
+    self.lastSaved = ko.computed(function() {
+        var request = self.lastSaveRequest();
+        if (request !== undefined) {
+            if (request.completedDate !== undefined) {
+                return request.completedDate.toGMTString();
+            } else {
+                return 'pending';
+            }
+        } else {
+            return 'never';
+        }
+    });
+
+    self.canRegister = ko.computed(function() {
+        var draft = self.draft();
+        return draft && draft.isApproved;
+    });
+
+    self.iterObject = $osf.iterObject;
+
+    // TODO: better extensions system?
+    self.extensions = {
+        'osf-upload': editorExtensions.Uploader
+    };
+};
+/**
+ * Load draft data into the editor
+ *
+ * @param {Draft} draft
+ **/
+RegistrationEditor.prototype.init = function(draft) {
+    var self = this;
+
+    self.draft(draft);
+    var metaSchema = draft.metaSchema;
+
+    var schemaData = {};
+    if (draft) {
+        schemaData = draft.schemaData || {};
+    }
+
+    // Set currentPage to the first page
+    self.currentPage(self.draft().pages()[0]);
+
+    self.serialized.subscribe(self.autoSave.bind(self));
+};
+/**
+ * @returns {Question[]} flat list of the current schema's questions
+ **/
+RegistrationEditor.prototype.flatQuestions = function() {
+    var self = this;
+
+    return self.draft().metaSchema.flatQuestions();
+};
+/**
+ * Creates a template context for editor type subtemplates. Looks for the data type
+ * in the extension map, and if found instantiates that type's ViewModel. Otherwise
+ * return the bare data Object
+ *
+ * @param {Object} data: data in current editor template scope
+ * @returns {Object|ViewModel}
+ **/
+RegistrationEditor.prototype.context = function(data) {
+    $.extend(data, {
+        save: this.save.bind(this),
+        readonly: this.readonly
+    });
+
+    if (this.extensions[data.type]) {
+        return new this.extensions[data.type](data);
```

Comment:
Note to self: extensions defined here. As far as I can tell there's only one extension. Are more planned?
https://github.com/CenterForOpenScience/osf.io/pull/4428/files#diff-c7726ca88c3b2e15d44ec996c513ec73R429